### DISCUSSION TITLE
Expo Router showing files in __test__ folders

### DISCUSSION THREAD

**Discussion by kyusungpark** [expo/router#889] [General]
Hi, I been writing tests and been putting those tests in __test folders in their respective locations where the file in located. I noticed that when I develop locally with expo I don't see the test files, but when I do my run my build, I see the test files as Tabs within the app.

I did try (__test__) but this did not work either, is there a way to hide these tabs from expo router?
I like to have them in the same folder as the other files instead of creating a separate folder just for the tests.

<img width="255" alt="Screenshot 2023-09-24 at 10 51 31 PM" src="https://github.com/expo/router/assets/58890841/fe83dfb2-1666-429a-b05b-1d9b77def838">
<img width="346" alt="Screenshot 2023-09-24 at 10 51 26 PM" src="https://github.com/expo/router/assets/58890841/85f25158-82c7-4126-9626-72803bdfed5c">

**Comment by marklawlor** [comment]
Expo Router doesn't support non-route files in the app directory.

**Reply by alexgshaw** [reply]
This feels like a pretty significant drawback. Is this feature on the roadmap?

**Reply by marklawlor** [reply]
No, it goes against our design philosophy. Co-locating route & non-route files makes sense in a directory based design (Svelte Kit, NextJS), but Expo Router uses a file based design

**Reply by alexgshaw** [reply]
Where do you recommend putting the components associated with a route? Having all components in a single directory is disorganized and splitting them into subfolders is inefficient because if you want to update a route name, you need to update it in two places.

Just curious what the recommended approach is in this case.

**Reply by marklawlor** [reply]
We recommend having a folder structure like so

```
/app
/components 
```

I don't understand your comment about having to update in two places, I'm assuming your talking about relative import paths? We recommend using path aliases to easily access components e.g `import "@/components/<my-components>"` and avoid relative import paths. The recent templates from Next.JS and Svelte Kit also adopt this pattern, as well as many starter templates from the community.

**Reply by alexgshaw** [reply]
Thanks for the quick response. That's not exactly what I'm referring to.

Here's an example of the dilemma I'm referring to. Say I have the following _desired_ folder structure.

### Desired Approach

```
app/
├── index.tsx
└── sign-in/
    ├── index.tsx
    └── components/
        └── sign-in-form.tsx
```

Well, with `expo-router` that is not possible. So the recommended approach would instead be one of these two options

### Option 1

This is the approach you are recommending IIUC.

```
app/
├── index.tsx
└── sign-in/
    └── index.tsx
components/
└── sign-in-form.tsx
```

The problem is that as the app grows, the `components` folder becomes increasingly unmanageable because of its flat structure (which is not the case in the desired approach). You could easily end up with 100+ components in one folder.

So maybe you take a different approach:

### Option 2

```
app/
├── index.tsx
└── sign-in/
    └── index.tsx
components/
└── sign-in/
    └── sign-in-form.tsx
```

This solves the flat `components` folder problem, but now whenever I want to change a route, I have to make edits in two places and update the imports in my route. In general, this is a lot of overhead and you run into the risk of your `components` and `app` directories falling out of sync.

Can you give me an example structure that doesn't run into those issues? If not, I think `expo-router` should seriously consider allowing non-route files inside of the `app` directory. Maybe the behavior could be something like "if no default component is exported, then don't render as a route" or have some prefix that can be put on a folder to not render it as a route, e.g. `^`.

(I'm a huge fan of the push for `expo-router` and directory-based routing, which is why I'm adding my two cents)

**Reply by marklawlor** [reply]
We recommend option 2.

I can understand your concerns but I believe your example is too simple and does not represent the typical structure of a native application. Most native applications will use [shared routes](https://docs.expo.dev/router/advanced/shared-routes/) (as opposed to a web application where they are not as common). When using shared routes the "route file" will exist in multiple places, not a single directory, so users will be using a "components" folder.

At this stage we're very unlikely to change to a completely different architecture. We know the pros/cons and the possible solutions - Expo Router has been in development for a while! But we like the direction we're currently moving and are unlikely to change.

Couple of last comments
- The downside of renaming files/imports is mostly offset modern editors, which will move a file and update all imports for you. If you find this frustrating this might be the solution for you.
- From my experience, keeping `/routes` and `/components` in sync ends up being an anti-pattern. Grouping components by functionality, not physical route name, works much better. This will often grow into something similar to [atomic design](https://atomicdesign.bradfrost.com/chapter-2/). But as multiple routes can display the same page, its not ideal to tie routes and pages together
- We are already uncomfortable with the number of "filename" rules and special exports that Expo Router has. Adding more rules just adds extra complexity and requires specialised knowledge. We're seeing other frameworks also move away from them, so this seems to be a shared sentiment across the community.
- We're not against co-location and plan to use them when appropriate. For example we're exploring [platform specific route files](https://github.com/expo/expo/pull/26047).

**Reply by khenkel-daxko** [reply]
Hi, I'm aware this discussion is a few months old already, but I'd like to add more thoughts here.
It's a common pattern to have test files next to their to-be-tested components, like this:
```
app
├── home
│   ├── user.test.tsx
│   └── user.tsx
├── index.test.tsx
├── index.tsx
├── sign-in.test.tsx
└── sign-in.tsx
```

With the current routing system Expo doesn't support this. My proposal is for this to be configurable in the **app.json** file. Next.js solves the same problem by making [pageExtensions](https://nextjs.org/docs/pages/api-reference/next-config-js/pageExtensions) configurable. This way you can define that all routes have to end with e.g. `.page.tsx` to be considered as a route, and everything else will be ignored.

**Reply by marklawlor** [reply]
To add another layer to this discussion, we are looking at supporting platform extensions for routes. This means you could have this structure

```
app
└── home
    ├── user.tsx
    ├── user.android.tsx
    ├── user.ios.tsx
    └── user.web.tsx
 ```
 
 If we added `.test` files, the structure will now be

```
app
└── home
    ├── user.tsx
    ├── user.test.tsx
    ├── user.android.tsx
    ├── user.android.test.tsx
    ├── user.ios.tsx
    ├── user.ios.test.tsx
    └── user.web.tsx
    └── user.web.test.tsx
 ```
 
I like co-location, but having eight files for a single route feels excessive. Also `.test.tsx` is not a standard in React Native. For example `jest-expo` use `-test.<platform>` to improve readability with platform extensions. This makes things harder because what if you have a route that ends in `-test`?
 
I'm a fan of Next.js, but the ecosystem in React Native is different, so concepts will not translate 1-1

**Reply by cameronolivier** [reply]
Adding to this - this makes sense for the simplest of routing/apps. As soon as you add any complexity you run into a potential nested horror story. 

Example:
A typical app would be using redux (or some state management). Along with Redux you might have style files, test files, route-specific components, route-specific utilities or constants, etc  - depending on the general approach one may have taken with FE architecture. 

For a given route, it's common to have:
```
app
└── home
    ├── components
    │   ├── userProfile.tsx
    │   ├── userProfile.test.tsx
    │   ├── userProfile.styles.tsx
    ├── index.tsx
    ├── home.test.tsx
    ├── home.reducer.tsx
    ├── home.middleware.tsx
    ├── home.reducer.tsx
    ├── home.styles.tsx
    └── home.selector.tsx
```
etc.

We've had routes with complex forms etc that need to be broken up (but are specific to that route).

I've found success with Tanstack Router (which is both directory and file-based) where one can wrap all non-route files away - but keep them collocated. 

```
app
└── home
    ├── -files
    │     ├── components
    │     │   ├── userProfile.tsx
    │     │   ├── userProfile.test.tsx
    │     │   ├── userProfile.styles.tsx
    │     ├── home.test.tsx
    │     ├── home.reducer.tsx
    │     ├── home.middleware.tsx
    │     ├── home.reducer.tsx
    │     ├── home.styles.tsx
    │     └── home.selector.tsx
    └── index.tsx
```
The alternative here would be to either run parallel directory structures (or some flattened version) or give up on collocation altogether and spread these files to the wind...

### Parallel
this quickly gets out of hand if you have deeply nested routes (excuse the forced example)
```
app
└── home
   └── user
   └── user-profile
   └── user-profile-edit
modules // or pages or routeFiles, etc
└── home
    │     ├── user
    │      │     ├── ....
    │     ├── user-profile
    │      │     ├── ....
    │     ├── components
    │     │   ├── userProfile.tsx
    │     │   ├── userProfile.test.tsx
    │     │   ├── userProfile.styles.tsx
    │     ├── home.test.tsx
    │     ├── home.reducer.tsx
    │     ├── home.middleware.tsx
    │     ├── home.reducer.tsx
    │     ├── home.styles.tsx
    │     └── home.selector.tsx
   ...
```
### To The Wind (grouped by _what_ not _where_):
```
app
└── home
   ├── user
   ├── user-profile
   └── user-profile-edit
reduces // or pages or routeFiles, etc
├──  home.reducer.ts
├── user.reducer.ts
└── ...
selectors
├──  home.reducer.ts
├── user.reducer.ts
└── ...
components
├──  routes
    └── home
       ├── user
       ├──user-profile
       ├──user-profile-edit
      └── ...
├──  common
    │     ├── ....
tests
.... 
utils
...
```

being able to group within routes - even at a higher level - would be supremely useful. We can already use (group) syntax, and <directory>/index.tsx as well - allowing you to effectively create a directory-based routing structure. It simply doesn't feel like including the ability to co-locate files specific to routes but which should not be routes is that far outside of the features already available (And would be super useful in simplifying the navigation of the overall directory structure in the app leading to better DX and onboarding). 

PS: Please excuse any issues with the directory structures - they were _tedious_ to handroll...

**Reply by anicioalexandre** [reply]
> Desired Approach
```
app/
├── index.tsx
└── sign-in/
    ├── index.tsx
    └── components/
        └── sign-in-form.tsx
```

In the `sign-in-form.tsx`, what if we avoid using a default export for the component? According to the Expo router documentation:

> A route file is defined by exporting a React component as the default value.

@marklawlor, do you think this approach could negatively affect performance during expo scanning process, especially in larger projects with many non-route files, or is the impact negligible?

**Reply by LovesWorking** [reply]
> marklawlor

@marklawlor @EvanBacon

Cameron Olivier makes a great point, and I completely agree. Developers should have the flexibility to structure their routes and organize their apps in the way that best suits their needs. Expo shouldn’t enforce an opinionated approach by labeling colocation as an anti-pattern, especially when many in the community see value in it.

React itself is built on the principle of unopinionated flexibility, allowing teams to structure their apps based on what works best for their specific needs. Expo should embrace this philosophy instead of dictating (x) approach.

If Expo Router isn't going to support colocation despite clear demand from the community, it's clear that Expo prioritizes its internal preferences over developer experience. Many frameworks have moved towards more flexible routing systems, why should Expo be the exception?

Instead of deciding what’s "right" for all apps and developers, why not give developers the option? That way, teams can work in a way that aligns with their preferences.
